### ISSUE TITLE
When client-side routing via Link, old route component is re-rendered twice before new route component is rendered

### ISSUE THREAD

**Issue by effulgentsia** [remix-run/remix#7755]
### What version of Remix are you using?

2.1.0

### Are all your remix dependencies & dev-dependencies using the same version?

- [X] Yes

### Steps to Reproduce

I created a demo here: https://github.com/effulgentsia/remix-demo-rerenders. This demo starts with your [Quick Start](https://remix.run/docs/en/main/start/quickstart), adds 2 routes and corresponsing `Link` tags to them from root.jsx, and adds loader functions to root.jsx and the 2 routes. It also adds `console.log()` statements to each of the components to see when they're executed.

[When](https://github.com/effulgentsia/remix-demo-rerenders/tree/5b8ae039795042dc7733fac2e77e89c4ed60d3b3) the components do not call `useLoaderData()`, then things work as I expect: if I refresh my browser on route1, clear the browser console, and then click the link to route2, only a single execution of route2 is logged in the browser console.

However, [when useLoaderData() is added](https://github.com/effulgentsia/remix-demo-rerenders/commit/8ba3c9e0b9a6a96d25856a9d7a437042af184089), then when I refresh my browser on route1, clear the browser console, and then click the link to route2, the following gets logged to the browser console:
```
Executing root.jsx's default export
Executing route1's default export
Executing root.jsx's default export
Executing route1's default export
Executing root.jsx's default export
Executing route2's default export
```

This is surprising. Why are all but the last line of the above triggered? By the way, this is all with the production build (`npx remix build; npx remix-serve build/index.js`). Each render is doubled when using the development build, but I think that's a separate known behavior of React development builds.

### Expected Behavior

Upon clicking the link to route2, I'd expect to see only the last line in the browser console, not the 5 before it.

### Actual Behavior

When the components don't call `useLoaderData()`, the actual result is what I expect above. But when the components do call `useLoaderData()`, I see the (seemingly unnecessary) 5 function calls above it (the root component getting re-rendered 3 times and the old route getting re-rendered twice).

**Comment by sergiodxa** [comment]
This is because useLoaderData reads from the Remix context and this one keeps track of the loading state so you can get it with useNavigation. Remember that [re-renders are not a bad things](https://kentcdodds.com/blog/fix-the-slow-render-before-you-fix-the-re-render) per se.

**Comment by brophdawg11** [comment]
I'm going to close this out since as @sergiodxa said, this is expected since Remix goes into a `loading` state during these navigations which causes the Remix context to update.  The reproduction repo seems to have been deleted as well so I couldn't look at the code to confirm.  If this is a different issue from the context rerenders, please recreate the reproduciton repo and we can dig in.

**Comment by effulgentsia** [comment]
> The reproduction repo seems to have been deleted

Sorry about that, I didn't notice that the repo defaulted to private when I created it. I just now made it public.

> this is expected since Remix goes into a loading state during these navigations which causes the Remix context to update

It's fine to keep this issue closed if this is a known and expected behavior. I can call `useMemo()` within components that do something expensive in their render. However, this seems to be a contradictory answer to the one in https://github.com/remix-run/remix/discussions/4182 unless there's a nuance here that I'm missing. It sounds like there's something in the context object that's changing that's causing the re-renders even if the data that the component is asking for from useLoaderData() isn't changing?
